internal/js_library/js_library.bzl

```diff
@@ -49,96 +53,124 @@ def write_amd_names_shim(actions, amd_names_shim, targets):
+           ) and
+           # exclude eg. external/npm/node_modules/protobufjs/node_modules/@types/node/index.d.ts
+           # these would be duplicates of the typings provided directly in another dependency.
+           # also exclude all /node_modules/typescript/lib/lib.*.d.ts files as these are determined by
+           # the tsconfig "lib" attribute
+           len(f.path.split("/node_modules/")) < 3 and f.path.find("/node_modules/typescript/lib/lib.") == -1
+    ])
+
+    transitive_declarations_depsets = [declarations]
+
+    for dep in ctx.attr.deps:
+        if DeclarationInfo in dep:
+            transitive_declarations_depsets.append(dep[DeclarationInfo].transitive_declarations)
+        if NpmPackageInfo in dep:
+            sources_depsets.append(dep[NpmPackageInfo].sources)
+        if JSModuleInfo in dep:
+            sources_depsets.append(dep[JSModuleInfo].sources)
+        if JSNamedModuleInfo in dep:
+            named_module_sources_depsets.append(dep[JSNamedModuleInfo].sources)
+
+    transitive_declarations = depset(transitive = transitive_declarations_depsets)
+    transitive_sources = depset(transitive = sources_depsets + named_module_sources_depsets)
 
     providers = [
         DefaultInfo(
-            files = files_depset,
-            runfiles = ctx.runfiles(files = ctx.files.srcs),
+            files = direct_sources,
+        ),
+        DeclarationInfo(
+            declarations = declarations,
+            transitive_declarations = transitive_declarations,
+            type_blacklisted_declarations = depset([]),
+        ),
+        js_module_info(
+            sources = direct_sources,
+            deps = ctx.attr.deps,
+        ),
+        js_named_module_info(
+            sources = direct_named_module_sources,
+            deps = ctx.attr.deps,
         ),
         AmdNamesInfo(names = ctx.attr.amd_names),
-        js_module_info(depset(js_files)),
     ]
 
     if ctx.attr.package_name:
         path = "/".join([p for p in [ctx.bin_dir.path, ctx.label.workspace_root, ctx.label.package] if p])
         providers.append(LinkablePackageInfo(
             package_name = ctx.attr.package_name,
             path = path,
-            files = files_depset,
+            files = depset([
+                transitive_sources,
+                transitive_declarations,
+            ]),
         ))
 
-    # Don't provide DeclarationInfo if there are no typings to provide.
-    # Improves error messaging downstream if DeclarationInfo is required.
-    if len(typings):
-        providers.append(declaration_info(depset(typings)))
+    if include_npm_package_info:
+        workspace_name = ctx.label.workspace_name if ctx.label.workspace_name else ctx.workspace_name
+        providers.append(NpmPackageInfo(
+            direct_sources = direct_sources,
+            sources = transitive_sources,
+            workspace = workspace_name,
+        ))
 
-    return providers
+    return struct(
+        typescript = struct(
```

Comment:
what happens if we don't provide the legacy string-typed "typescript" provider? if something is broken we should maybe update that use site rather than carry this forward, as it will be a breaking change to remove it again

Reply:
I've carried this from the old `node_module_library`. But now that I think better about it, I think it's safe to remove it